# Unlocking Data Insights: A SQL Case Study on the Northwind Database


This project explores sales performance, customer behavior, and employee efficiency using SQL queries on the Northwind database. The goal is to demonstrate practical SQL skills for business insights.


## 🎯 Project Objectives

This project aims to analyze the Northwind dataset using SQL to uncover key business insights. The specific objectives are to:

- Evaluate sales performance across products, categories, and regions.

- Identify top-performing employees based on total sales and order volume.

- Examine customer purchasing behavior and highlight the most valuable clients.

- Track order and shipment trends to understand delivery efficiency.

- Generate data-driven insights that can guide decision-making in similar retail or trading contexts.

## 📊 Dataset Overview

The **Northwind dataset** represents a fictional trading company that sells various products to customers worldwide. It contains relational tables that capture key aspects of the company’s operations, including orders, shipments, and employee activities.  

You can access the dataset here: [Northwind Sample Database)](https://mavenanalytics.io/)  

The main tables include:  
- **Customers:** Information about clients and their contact details.  
- **Products:** Details of items sold, including unit prices and stock levels.  
- **Product Category:** Groups products into related categories.  
- **Orders:** Records of customer purchases handled by employees.  
- **Order_Details:** Links each order to the specific products purchased and quantities sold.  
- **Employees:** Information about staff responsible for processing orders.  
- **Shippers:** Companies responsible for delivering customer orders.  

These interconnected tables make it possible to analyze **sales trends, product performance, and employee efficiency** across the company’s operations.


## Entity Relationship (ER) Diagram

The ER diagram below illustrates how the main entities in the Northwind database are connected through relationships.

![Northwind ER Diagram](northwind_er_diagram.png)


## Database Setup

The Northwind dataset was imported into **MySQL Workbench**, where all analysis queries were executed. To connect Jupyter Notebook to the MySQL database, we use the `mysql.connector` library in Python. This allows SQL queries to be written and run directly from the notebook, enabling seamless integration between data retrieval and analysis.


## Connecting to our MySQL database

In [1]:
!pip install -q ipython-sql mysqlclient python-dotenv

In [2]:
%load_ext sql

In [3]:
from dotenv import load_dotenv
import os
from pathlib import Path
import urllib.parse  # needed for encoding

# 1️⃣ Load .env file
dotenv_path = Path('.') / '.env'
load_dotenv(dotenv_path=dotenv_path)

# 2️⃣ Get environment variables
MYSQL_USER = os.getenv("MYSQL_USER")
MYSQL_PASSWORD = os.getenv("MYSQL_PASSWORD")  # raw password from .env
MYSQL_HOST = os.getenv("MYSQL_HOST")
MYSQL_DATABASE = os.getenv("MYSQL_DATABASE")

# 3️⃣ Encode the password (this step must come before using it in connection string)
MYSQL_PASSWORD = urllib.parse.quote(MYSQL_PASSWORD)

# 4️⃣ Build connection string for ipython-sql
connection_string = f"mysql://{MYSQL_USER}:{MYSQL_PASSWORD}@{MYSQL_HOST}/{MYSQL_DATABASE}"

# 5️⃣ Connect using %%sql
%sql $connection_string

'Connected: root@northwind_traders'

## Analyzing Data in the Northwind Traders Database

After successfully connecting to our MySQL database, we can begin performing analytical queries to uncover meaningful business insights from the **Northwind Traders** dataset. This analysis focuses on understanding customer distribution, product performance, employee activity, and overall sales patterns.  

Using SQL, we’ll generate summaries and aggregations that reveal how different parts of the business contribute to revenue and operations.


### 1. Customer Insights & Sales Performance:

### Business Question:  
Management is interested in identifying the **top 5 customers globally** who generate the **highest revenue**, along with their **respective countries and cities**.


In [5]:
%%sql
WITH customer_revenue AS (
   SELECT
      cus.customerID,
      cus.companyName,
      cus.country,
      cus.city,
      ROUND(SUM(ordet.unitPrice * ordet.quantity * (1 - ordet.Discount)), 2) 
     AS total_revenue
   FROM
      customers AS cus
   LEFT JOIN 
      orders AS ord
    ON 
      cus.customerID = ord.customerID
   LEFT JOIN 
      order_details AS ordet
    ON 
      ord.orderID = ordet.orderID
   GROUP BY
       1,2,3,4
)

SELECT
 customerID,
 companyName,
 country,
 city,
 total_revenue
FROM
 customer_revenue
ORDER BY
 total_revenue DESC
LIMIT 5;

 * mysql://root:***@127.0.0.1:3306/northwind_traders
5 rows affected.


[('QUICK', 'QUICK-Stop', 'Germany', 'Cunewalde', 117483.39),
 ('SAVEA', 'Save-a-lot Markets', 'USA', 'Boise', 115673.39),
 ('ERNSH', 'Ernst Handel', 'Austria', 'Graz', 113236.68),
 ('HUNGO', 'Hungry Owl All-Night Grocers', 'Ireland', 'Cork', 57317.39),
 ('RATTC', 'Rattlesnake Canyon Grocery', 'USA', 'Albuquerque', 52245.9)]

This result shows the top 5 customers globally ranked by total revenue. It highlights which companies contribute most to sales performance and shows their locations (country and city), enabling management to see where key revenue drivers are concentrated.

### Business Question:

Concerns have been raised about **customer retention**. There is a need to identify customers who have placed only a few orders or have not made purchases within the 180 days leading up to May 6, 2015, as these customers may be at **risk of churn**.

In [8]:
%%sql
WITH customer_churn AS (
   SELECT
      cus.customerID,
      cus.companyName,
      COUNT(ord.orderID) AS order_count,
      MAX(ord.orderDate) AS last_order_date,
   CASE 
       WHEN MAX(ord.orderDate) IS NULL
        OR MAX(ord.orderDate) < DATE('2015-05-06') - INTERVAL 180 DAY
        OR COUNT(ord.orderID) <= 1
       THEN 'At Risk of Churn'
        ELSE 'Active'
      END AS churn_status
   FROM
      customers AS cus
   LEFT JOIN
      orders AS ord
   ON cus.customerID = ord.customerID
   GROUP BY
      1, 2
   ORDER BY
      last_order_date ASC
)

SELECT 
 *
FROM
 customer_churn
WHERE
 churn_status = 'At Risk of Churn';

 * mysql://root:***@127.0.0.1:3306/northwind_traders
7 rows affected.


[('FISSA', 'FISSA Fabrica Inter. Salchichas S.A.', 0, None, 'At Risk of Churn'),
 ('PARIS', 'Paris spécialités', 0, None, 'At Risk of Churn'),
 ('CENTC', 'Centro comercial Moctezuma', 1, datetime.date(2013, 7, 18), 'At Risk of Churn'),
 ('LAZYK', 'Lazy K Kountry Store', 2, datetime.date(2014, 5, 22), 'At Risk of Churn'),
 ('HUNGC', 'Hungry Coyote Import Store', 5, datetime.date(2014, 9, 8), 'At Risk of Churn'),
 ('MEREP', 'Mère Paillarde', 13, datetime.date(2014, 10, 30), 'At Risk of Churn'),
 ('FAMIA', 'Familia Arquibaldo', 7, datetime.date(2014, 10, 31), 'At Risk of Churn')]

This result shows a list of customers identified as being at risk of churn based on low order frequency or inactivity within the last 180 days before May 6, 2015.

### 2. Product Performance & Category Insights:

### Business Question:

The product team needs to know which **product categories** contribute most to **overall revenue and profitability**.

In [9]:
%%sql
SELECT 
  cat.categoryID,
  cat.categoryName,
  ROUND(SUM(ord.unitPrice * ord.quantity), 2) AS total_revenue
FROM
  products AS pro
LEFT JOIN
  order_details AS ord
ON
  pro.productID = ord.productID
LEFT JOIN
  categories AS cat
ON
  pro.categoryID = cat.categoryID
GROUP BY
  1,2
ORDER BY
  3 DESC;

 * mysql://root:***@127.0.0.1:3306/northwind_traders
8 rows affected.


[(1, 'Beverages', 286526.95),
 (4, 'Dairy Products', 251330.5),
 (6, 'Meat & Poultry', 178188.8),
 (3, 'Confections', 177099.1),
 (8, 'Seafood', 141623.09),
 (2, 'Condiments', 113694.75),
 (7, 'Produce', 105268.6),
 (5, 'Grains & Cereals', 100726.8)]

This result shows the **total revenue** contributed by each of the **eight product categories**. **Beverages (categoryID = 1)** stands out as the **leading contributor to overall revenue**.

### Business Question:

**Inventory optimization** is critical. Management wants to identify **products with consistently low sales volumes**, defined as **those with fewer than 100 total units sold**, that may be **considered for discontinuation**.

In [10]:
%%sql
WITH prod_stats AS (    
  SELECT 
      p.productID,
      p.productName,
      SUM(od.quantity) AS total_units_sold,
      ROUND(SUM(od.quantity * od.unitPrice), 2) AS total_revenue,
      CASE 
          WHEN SUM(od.quantity) < 100 THEN 'Low Sales'
          WHEN SUM(od.quantity) BETWEEN 100 AND 500 
        THEN 'Medium Sales'
          ELSE 'High Sales'
      END AS sales_category,
    
    CASE 
        WHEN SUM(od.quantity) < 100 
      THEN 'Consider for Discontinuation'
        ELSE 'Keep Product'
    END AS product_status

  FROM 
   products AS p
  JOIN 
   order_details AS od 
  ON 
   p.productID = od.productID
  GROUP BY 
   p.productID, p.productName
  ORDER BY 
   total_units_sold ASC
)
    
SELECT
     *
FROM
     prod_stats
WHERE 
     sales_category = 'Low Sales'
       AND product_status = 'Consider for Discontinuation';

 * mysql://root:***@127.0.0.1:3306/northwind_traders
1 rows affected.


[(9, 'Mishi Kobe Niku', Decimal('95'), 8827.0, 'Low Sales', 'Consider for Discontinuation')]

The result highlights **Mishi Kobe Niku (ProductID: 9)** as the product with the **lowest total units sold**, indicating that it should be **considered for discontinuation**.

### 3. Shipping & Delivery Performance:

### Business Question: 

Operations are evaluating **shipper performance**. Which **shipping companies** handles the **highest number of orders**, and what is their **average delivery time?**

In [11]:
%%sql
WITH shipper_performance AS (
    SELECT
       shipperID,
       COUNT(orderID) AS number_of_orders,
       AVG(DATEDIFF(shippedDate, orderDate)) AS avg_delivery_time
    FROM
       orders
    GROUP BY
        shipperID
)
    
SELECT
    ship_p.shipperID,
    companyName,
    number_of_orders,
    avg_delivery_time
FROM
 shippers AS ship
LEFT JOIN 
 shipper_performance AS ship_p
ON
 ship.shipperID = ship_p.shipperID
ORDER BY 
 number_of_orders DESC;

 * mysql://root:***@127.0.0.1:3306/northwind_traders
3 rows affected.


[(2, 'United Package', 326, Decimal('9.2349')),
 (3, 'Federal Shipping', 255, Decimal('7.4739')),
 (1, 'Speedy Express', 249, Decimal('8.5714'))]

The result shows the order distribution and average delivery time by shipper. **Federal Shipping** delivers the fastest on average, while **United Package handles the highest number of orders but with the slowest delivery times**.

### Business Question:

Management also wants to assess **shipping delays**. Which shippers handle the most orders, and how do they compare in terms of average delivery times and the share of orders exceeding the **7-day delivery benchmark?**

In [12]:
%%sql
SELECT
    ship.shipperID,
    ship.companyName,
    COUNT(orderID) AS total_orders,
    AVG(DATEDIFF(shippedDate, orderDate)) AS avg_delivery_time,
    SUM(CASE 
          WHEN DATEDIFF(shippedDate, orderDate) > 7 
         THEN 1 ELSE 0 END) 
       AS delayed_orders,
    ROUND(100.0 * SUM(CASE 
                         WHEN DATEDIFF(shippedDate, orderDate) > 7 
                       THEN 1 ELSE 0 END) / COUNT(orderID), 2) 
       AS delay_rate
FROM
    orders AS ord
LEFT JOIN
 shippers AS ship
ON
 ord.shipperID = ship.shipperID
GROUP BY
    shipperID, companyName
ORDER BY
    delay_rate DESC;

 * mysql://root:***@127.0.0.1:3306/northwind_traders
3 rows affected.


[(2, 'United Package', 326, Decimal('9.2349'), Decimal('141'), Decimal('43.25')),
 (1, 'Speedy Express', 249, Decimal('8.5714'), Decimal('102'), Decimal('40.96')),
 (3, 'Federal Shipping', 255, Decimal('7.4739'), Decimal('93'), Decimal('36.47'))]

This result highlights clear differences in shipper performance. **Federal Shipping demonstrates the quickest turnaround times**, delivering orders faster than its competitors. In contrast, **United Package handles the largest share of orders but also records the highest delay rate**, suggesting a trade-off between volume and reliability.

### 4. Employee Productivity & Operations Efficiency:

### Business Question: 

HR and operations want to identify **employees who process the highest number of orders**, along with their **average order value**.

In [13]:
%%sql
WITH ord_val AS (
 SELECT
        ord.employeeID,
        COUNT(DISTINCT ords.orderID) AS no_of_orders,
        SUM(ords.unitPrice * ords.quantity) AS total_sales
 FROM 
  orders AS ord
 LEFT JOIN
  order_details AS ords
 ON
  ord.orderID = ords.orderID
 GROUP BY
  1
)    
SELECT
   emp.employeeID,
    employeeName,
    no_of_orders,
    ROUND(ov.total_sales / ov.no_of_orders, 2) AS avg_ord_value
FROM
    ord_val AS ov
LEFT JOIN
 employees AS emp
ON
 ov.employeeID = emp.employeeID
ORDER BY
 3 DESC;

 * mysql://root:***@127.0.0.1:3306/northwind_traders
9 rows affected.


[(4, 'Margaret Peacock', 156, 1603.77),
 (3, 'Janet Leverling', 127, 1677.57),
 (1, 'Nancy Davolio', 123, 1643.44),
 (8, 'Laura Callahan', 104, 1281.74),
 (2, 'Andrew Fuller', 96, 1851.55),
 (7, 'Robert King', 72, 1962.44),
 (6, 'Michael Suyama', 67, 1167.14),
 (9, 'Anne Dodsworth', 43, 1929.4),
 (5, 'Steven Buchanan', 42, 1799.23)]

This result highlights **employee order performance**, showing **Margaret Peacock processed the highest number of orders (156) with a strong average value**, while **Robert King, despite handling fewer orders, achieved the highest average order value at 1,962.44**. The table illustrates how both order volume and order value vary among employees.

### Business Question: 

There is interest in **measuring employee efficiency**. How long does it typically take employees to process orders from placement to shipment?

In [14]:
%%sql
SELECT
    emp.employeeID,
    emp.employeeName,
    AVG(DATEDIFF(ord.shippedDate, ord.orderDate)) AS avg_processing_days,
    MIN(DATEDIFF(ord.shippedDate, ord.orderDate)) AS fastest_days,
    MAX(DATEDIFF(ord.shippedDate, ord.orderDate)) AS slowest_days,
    COUNT(ord.orderID) AS total_orders
FROM
    employees AS emp
INNER JOIN
    orders AS ord
ON
    emp.employeeID = ord.employeeID
GROUP BY
    emp.employeeID, emp.employeeName
ORDER BY
    avg_processing_days ASC;

 * mysql://root:***@127.0.0.1:3306/northwind_traders
9 rows affected.


[(5, 'Steven Buchanan', Decimal('7.0238'), 1, 25, 42),
 (1, 'Nancy Davolio', Decimal('7.7583'), 1, 34, 123),
 (2, 'Andrew Fuller', Decimal('8.0538'), 1, 32, 96),
 (7, 'Robert King', Decimal('8.3768'), 1, 37, 72),
 (3, 'Janet Leverling', Decimal('8.4252'), 1, 35, 127),
 (8, 'Laura Callahan', Decimal('8.6800'), 1, 37, 104),
 (4, 'Margaret Peacock', Decimal('8.8212'), 1, 35, 156),
 (6, 'Michael Suyama', Decimal('9.0923'), 1, 32, 67),
 (9, 'Anne Dodsworth', Decimal('10.8571'), 1, 34, 43)]

This result shows the **overall efficiency of employees in processing orders**, with average turnaround times generally ranging between **7 and 11 days**. Most employees fall within the 7–9 day range, indicating consistent performance across the team. **Steven Buchanan achieved the fastest average turnaround of 7.02 days**, while **Anne Dodsworth recorded the slowest at 10.86 days**.

### 5. Sales Trend & Strategic Insights:

### Business Question: 

The strategy team is examining seasonality. Are there particular months or periods with peak sales activity?

In [15]:
%%sql
SELECT
 Month,
    Year,
    ROUND(SUM(Sales), 2) AS Total_Sales
FROM
     (SELECT 
        MONTHNAME(orderDate) AS Month,
        YEAR(orderDate) AS Year,
        ROUND((unitPrice * quantity), 2) AS sales
     FROM
        orders
     LEFT JOIN
        order_details
     ON 
        orders.orderID = order_details.orderID
     ORDER BY 
        2 ASC
              ) AS sub_sales
GROUP BY
 1,2
ORDER BY
 Total_sales DESC
LIMIT
 5;

 * mysql://root:***@127.0.0.1:3306/northwind_traders
5 rows affected.


[('April', 2015, 134630.56),
 ('March', 2015, 109825.45),
 ('February', 2015, 104561.95),
 ('January', 2015, 100854.72),
 ('December', 2014, 77476.26)]

This result shows the monthly sales trend, with steady growth from December 2014 to April 2015, where **sales peaked at 134,630.56 in April 2015**.

### Business Question: 

Management wants to evaluate **year-over-year growth**. How have yearly sales changed over time in regions such as the **USA, UK, and Belgium,** and what is the **year-over-year growth compared to the previous year?**

In [16]:
%%sql
WITH sales_cte AS (
   SELECT
      country,
      year,
      sales,
     LAG(sales,1) OVER (
                       PARTITION BY country
                        ORDER BY year) AS Previous_sales
   FROM   
        (SELECT
             country,
             YEAR(orderDate) AS year,
             ROUND(SUM(ROUND((unitprice * quantity),2)),2) AS sales
        FROM
             orders AS ord
        LEFT JOIN
             customers AS cus
        ON
             ord.customerID = cus.customerID
        LEFT JOIN
             order_details AS ordet
        ON
             ord.orderID = ordet.orderID
        GROUP BY
              1, 2
        ORDER BY
               3 DESC
                       ) AS sub_salestrend
)
SELECT
    country,
    year,
    sales,
    IFNULL(Previous_sales, 0) AS previous_sales,
    CASE
          WHEN previous_sales IS NULL THEN 0 
       ELSE ROUND((sales - Previous_sales),2) 
   END AS sales_diff_over_year
FROM
 sales_cte
WHERE country IN 
 ('Belgium', 'USA', 'UK')
ORDER BY
 country, year;

 * mysql://root:***@127.0.0.1:3306/northwind_traders
9 rows affected.


[('Belgium', 2013, 6438.8, 0.0, 0.0),
 ('Belgium', 2014, 12087.1, 6438.8, 5648.3),
 ('Belgium', 2015, 16609.08, 12087.1, 4521.98),
 ('UK', 2013, 9654.0, 0.0, 0.0),
 ('UK', 2014, 27832.6, 9654.0, 18178.6),
 ('UK', 2015, 23129.91, 27832.6, -4702.69),
 ('USA', 2013, 41907.8, 0.0, 0.0),
 ('USA', 2014, 121037.7, 41907.8, 79129.9),
 ('USA', 2015, 100621.48, 121037.7, -20416.22)]

This result shows the **year-over-year sales trend for Belgium, the UK, and the USA between 2013 and 2015**, highlighting strong growth in Belgium and the USA from 2013 to 2014, followed by a sales decline in the UK and USA in 2015.

## Conclusion & Recommendation:


The analysis highlights key business dynamics. Revenue is concentrated among a few top customers, while others show low engagement or churn risk. Product performance varies — categories like Beverages drive strong sales, whereas items like Mishi Kobe Niku have limited demand. Shipping efficiency is mixed: Federal Shipping is fastest, but United Package handles larger volumes with more delays. Employee performance is mostly consistent, with a few excelling in speed or workload. Regional and seasonal patterns show strong growth in Belgium and the USA, but slower momentum in the UK.

These findings suggest actionable steps: strengthen ties with top customers, re-engage inactive ones, and focus resources on high-performing products. Improve shipping balance by optimizing carrier assignments, support employee development based on top performers, and plan around seasonal and regional trends to sustain long-term growth.